### ISSUE TITLE
adolc tests are broken

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#14582]
The MPI and serial Jenkins build shows several (all?) adolc tests failing with
```
An error occurred in line <1621> of file </jenkins/workspace/dealii-mpi_PR-14568/source/differentiation/ad/ad_drivers.cc> in function
    std::enable_if_t<(dealii::Differentiation::AD::ADNumberTraits<ADNumberType>::type_code == dealii::Differentiation::AD::NumberTypes::adolc_tapeless)> dealii::Differentiation::AD::internal::configure_tapeless_mode(unsigned int) [with ADNumberType = adtl::adouble; std::enable_if_t<(dealii::Differentiation::AD::ADNumberTraits<ADNumberType>::type_code == dealii::Differentiation::AD::NumberTypes::adolc_tapeless)> = void]
The violated condition was: 
    n_live_variables == 0
```

see #14568 and #14576
and https://ci.tjhei.info/job/dealii-mpi/job/PR-14568/1/testReport/junit/projectroot.tests.adolc.helper_scalar_coupled_3_components_01_2debug/mpirun=1/adolc_helper_scalar_coupled_3_components_01_2_mpirun_1_debug/

I don't know enough to understand what is going on here...

**Comment by tamiko** [comment]
@tjhei Do you build adolc via candi? For reference, I have (system) AdolC enabled for Ubuntu 20.04 and 22.04: https://cdash.dealii.org/build/588 https://cdash.dealii.org/build/585 and I don't see such fallout :-|

**Comment by tamiko** [comment]
@tjhei Can I try to reproduce with a docker image? :-)

**Comment by tjhei** [comment]
Yes,candi: https://github.com/dealii/candi/blob/dealii-9.4/deal.II-toolchain/packages/adolc.package

> Can I try to reproduce with a docker image? :-)
Sure, this should work:
https://github.com/dealii/dealii/blob/e4382e01ae2219e02c00a28f07a710398c92d7cc/contrib/ci/Jenkinsfile.serial#L48
https://github.com/dealii/dealii/blob/e4382e01ae2219e02c00a28f07a710398c92d7cc/contrib/ci/Jenkinsfile.serial#L98-L114

**Comment by jppelteret** [comment]
This failure is triggered by the fact that we have the `--enable-traceless_refcounting` flag [set here](https://github.com/dealii/candi/blob/dealii-9.4/deal.II-toolchain/packages/adolc.package#L37). IIRC this option is new to ADOL-C 2.7.2, and I think that most package managers to not activate it by default.

I tried to reproduce this error locally, but since I'm using the docker image I can't be sure of how ADOL-C is being built. I also had a look through the code and do not see anything obviously wrong. So I'll have build it externally to try to debug the issue. 

What confuses me is that I am quite certain that I developed [this code](https://github.com/dealii/dealii/blob/master/source/differentiation/ad/ad_drivers.cc#L1587-L1643) with the most conservative ADOL-C build options active. Anyway, I'll try to look into this over the holiday period.
